File: app/src/lib/models/monitor.ts
import type { ListResult, RecordModel } from 'pocketbase';
import { z } from 'zod';
import { type EntityIdType } from '$lib/constants';
import { type IBaseEntity, BaseEntityList, BaseEntityValidator, PaginationData } from './';

export const MonitorTypes = ['website', 'ping', 'port', 'heartbeat', 'duplicati', 'dns'] as const;
export type MonitorType = (typeof MonitorTypes)[number];

// The monitor types the worker can actually run today. The others exist in the
// enum (schema + roadmap) but must not be creatable until their worker
// implementation lands — otherwise a user can create a monitor that only ever
// reports "unsupported monitor type" (issue #82). This is the single source of
// truth: as each type ships, add it here and it re-enables in both the form
// (selectable) and the API (accepted by the validator below).
export const SupportedMonitorTypes = [
	'website',
	'port',
	'ping',
	'heartbeat',
	'dns'
] as const satisfies readonly MonitorType[];

export function isSupportedMonitorType(type: string): boolean {
	return (SupportedMonitorTypes as readonly string[]).includes(type);
}

// Parse a "Name: Value" per-line headers textarea into a map (blank lines and
// lines without a colon are ignored).
export function parseHeaders(raw: string | undefined): Record<string, string> {
	const out: Record<string, string> = {};
	for (const line of (raw ?? '').split('\n')) {
		const i = line.indexOf(':');
		if (i <= 0) continue;
		const k = line.slice(0, i).trim();
		const v = line.slice(i + 1).trim();
		if (k) out[k] = v;
	}
	return out;
}

// Render a headers map back into the "Name: Value" per-line textarea form.
export function headersToLines(headers: Record<string, string> | undefined | null): string {
	if (!headers) return '';
	return Object.entries(headers)
		.map(([k, v]) => `${k}: ${v}`)
		.join('\n');
}

function toPositiveInt(value: any): number | null {
	const n = Number(value);
	return Number.isFinite(n) && n > 0 ? Math.trunc(n) : null;
}

// Assemble the monitor `config` JSON from the flat form fields, or null when
// nothing non-default is set. Every field is omitted at its default so an unset
// config keeps the original behaviour (GET, 200–399 = up, follow redirects).
export function toMonitorConfig(item: any): Record<string, unknown> | null {
	const cfg: Record<string, unknown> = {};

	if (item.keywordMode === 'contains' || item.keywordMode === 'absent') {
		cfg.keyword = (item.keyword ?? '').trim();
		cfg.keywordMode = item.keywordMode;
	}

	const method = (item.method ?? '').toString().toUpperCase();
	if (method && method !== 'GET') cfg.method = method;

	const headers = parseHeaders(item.headers);
	if (Object.keys(headers).length) cfg.headers = headers;

	const expected = toPositiveInt(item.expectedStatus);
	if (expected) cfg.expectedStatus = expected;

	// Default is to follow; only persist the opt-out.
	if (item.followRedirects === false || item.followRedirects === 'false')
		cfg.followRedirects = false;

	const timeout = toPositiveInt(item.timeoutSecs);
	if (timeout) cfg.timeoutSecs = timeout;

	const windows = toWindows(item.maintenanceWindows);
	if (windows.length) cfg.maintenanceWindows = windows;

	const latency = toPositiveInt(item.latencyThresholdMs);
	if (latency) cfg.latencyThresholdMs = latency;

	// DNS options (only persisted for dns monitors, and only when non-default).
	const dnsRecordType = (item.dnsRecordType ?? '').toString().trim().toUpperCase();
	if (dnsRecordType && dnsRecordType !== 'A') cfg.dnsRecordType = dnsRecordType;
	const dnsExpectedValue = (item.dnsExpectedValue ?? '').toString().trim();
	if (dnsExpectedValue) cfg.dnsExpectedValue = dnsExpectedValue;
	const dnsResolver = (item.dnsResolver ?? '').toString().trim();
	if (dnsResolver) cfg.dnsResolver = dnsResolver;

	return Object.keys(cfg).length ? cfg : null;
}

export const MonitorStatuses = ['up', 'down', 'pending', 'paused'] as const;
export type MonitorStatus = (typeof MonitorStatuses)[number];

const schema = z
	.object({
		name: z.string().min(1, 'Name is required'),
		type: z.enum(MonitorTypes).refine(isSupportedMonitorType, {
			message: 'This monitor type is not available yet'
		}),
		target: z.string().optional().default(''),
		// Must match infrastructure/pb_schema.json's monitors.interval min and
		// corelib's models.MinIntervalSeconds (#319). If this floor is lower than
		// the schema's, the API accepts an interval PocketBase then rejects, and
		// the user gets an opaque 500 instead of this field message.
		interval: z.coerce.number().int().min(30, 'Interval must be at least 30 seconds'),
		keyword: z.string().optional(),
		keywordMode: z.string().optional(),
		method: z.string().optional(),
		headers: z.string().optional(),
		expectedStatus: z.union([z.number(), z.null()]).optional(),
		followRedirects: z.union([z.boolean(), z.string()]).optional(),
		timeoutSecs: z.union([z.number(), z.null()]).optional(),
		maintenanceWindows: z.array(z.any()).optional(),
		latencyThresholdMs: z.union([z.number(), z.null()]).optional(),
		dnsRecordType: z.string().optional(),
		dnsExpectedValue: z.string().optional(),
		dnsResolver: z.string().optional()
	})
	// A `port` monitor is a TCP connect, so it needs an explicit host:port —
	// otherwise the worker can't dial it and the monitor would sit down forever.
	.superRefine((val, ctx) => {
		// Probed types need a target to reach; a heartbeat is reached the other
		// way (the job checks in), so it has none.
		if (val.type !== 'heartbeat' && !(val.target ?? '').trim()) {
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				path: ['target'],
				message: 'Target is required'
			});
		}
		if (val.type === 'port') {
			const m = /^(.+):(\d{1,5})$/.exec((val.target ?? '').trim());
			const port = m ? Number(m[2]) : NaN;
			if (!m || port < 1 || port > 65535) {
				ctx.addIssue({
					code: z.ZodIssueCode.custom,
					path: ['target'],
					message: 'Enter host:port (e.g. example.com:443)'
				});
			}
		}
		// A body assertion needs the text to look for.
		if (
			(val.keywordMode === 'contains' || val.keywordMode === 'absent') &&
			!(val.keyword ?? '').trim()
		) {
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				path: ['keyword'],
				message: 'Enter the text to check for'
			});
		}
		if (val.expectedStatus != null && (val.expectedStatus < 100 || val.expectedStatus > 599)) {
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				path: ['expectedStatus'],
				message: 'Status code must be 100–599'
			});
		}
		if (val.timeoutSecs != null && (val.timeoutSecs < 1 || val.timeoutSecs > 120)) {
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				path: ['timeoutSecs'],
				message: 'Timeout must be 1–120 seconds'
			});
		}
		if (val.latencyThresholdMs != null && val.latencyThresholdMs < 1) {
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				path: ['latencyThresholdMs'],
				message: 'Threshold must be a positive number of milliseconds'
			});
		}
	});

export type IMonitorItem = z.infer<typeof schema>;

function toInterval(value: any): number {
	const n = Number(value);
	return Number.isFinite(n) ? Math.trunc(n) : 60;
}

function toBool(value: any): boolean {
	return value === true || value === 'true' || value === 1 || value === '1';
}

function toZones(value: any): string[] {
	if (Array.isArray(value)) return value.filter(Boolean);
	if (typeof value === 'string')
		return value
			.split(',')
			.map((z) => z.trim())
			.filter(Boolean);
	return [];
}

function toNumberOrNull(value: any): number | null {
	if (value === undefined || value === null || value === '') return null;
	const n = Number(value);
	return Number.isFinite(n) ? n : null;
}

export type MaintenanceWindow = { start: string; end: string };

// Accept maintenance windows as an array (API record) or a JSON string (round-
// tripped through FormData); keep only entries with both a start and an end.
export function toWindows(value: any): MaintenanceWindow[] {
	let arr = value;
	if (typeof value === 'string' && value.trim()) {
		try {
			arr = JSON.parse(value);
		} catch {
			return [];
		}
	}
	if (!Array.isArray(arr)) return [];
	return arr.filter((w) => w && w.start && w.end).map((w) => ({ start: w.start, end: w.end }));
}

// Read the HTTP-option fields from either an API record (nested `config` object)
// or flat form fields, defaulting each to its original behaviour when unset.
function configFields(item: any) {
	const c = item.config;
	return {
		keyword: c?.keyword ?? item.keyword ?? '',
		keywordMode: c?.keywordMode ?? item.keywordMode ?? '',
		method: c?.method ?? item.method ?? 'GET',
		headers: c?.headers ? headersToLines(c.headers) : (item.headers ?? ''),
		expectedStatus: toNumberOrNull(c?.expectedStatus ?? item.expectedStatus),
		followRedirects: c
			? (c.followRedirects ?? true)
			: item.followRedirects !== undefined
				? toBool(item.followRedirects)
				: true,
		timeoutSecs: toNumberOrNull(c?.timeoutSecs ?? item.timeoutSecs),
		maintenanceWindows: toWindows(c?.maintenanceWindows ?? item.maintenanceWindows),
		latencyThresholdMs: toNumberOrNull(c?.latencyThresholdMs ?? item.latencyThresholdMs),
		dnsRecordType: c?.dnsRecordType ?? item.dnsRecordType ?? 'A',
		dnsExpectedValue: c?.dnsExpectedValue ?? item.dnsExpectedValue ?? '',
		dnsResolver: c?.dnsResolver ?? item.dnsResolver ?? ''
	};
}

export class MonitorItem implements IBaseEntity {
	id: EntityIdType;
	name: string;
	type: MonitorType;
	target: string;
	interval: number;
	enabled: boolean;
	zones: string[];
	tags: string[];
	escalationPolicy: string;
	// Per-monitor HTTP options (stored in the monitor `config` JSON).
	keyword: string;
	keywordMode: string;
	method: string;
	headers: string;
	expectedStatus: number | null;
	followRedirects: boolean;
	timeoutSecs: number | null;
	maintenanceWindows: MaintenanceWindow[];
	latencyThresholdMs: number | null;
	dnsRecordType: string;
	dnsExpectedValue: string;
	dnsResolver: string;
	// Read-only display fields (written by the worker / evaluator, enriched by the API).
	status: MonitorStatus;
	// What the evaluator's last verdict was actually taken on (#328): the zones
	// that took part, and the subset that had a fresh check and therefore voted.
	// Read-only, and deliberately not re-derived here — freshness depends on the
	// monitor's effective interval inside the evaluator, so any second guess
	// drifts from the decision that was really made.
	consensusZones: string[];
	consensusFresh: string[];
	lastChecked: string | null;
	uptime24h: number | null;
	certExpiresAt: string | null;
	// When the target's DOMAIN registration lapses — an infrequent, cached
	// RDAP/WHOIS lookup the evaluator runs (distinct from the TLS cert above).
	// Null when unknown (no data, non-domain target, or not yet looked up).
	domainExpiresAt: string | null;
	// When the monitor most recently went down (latest incident's start), or null if
	// it has no incidents on record. Enriched by the list API from the `incidents`
	// collection (#279).
	lastDowntime: string | null;
	// Heartbeat check-in token (empty for probed types); used to build the /ping URL.
	token: string;

	constructor(data: FormData | RecordModel | any) {
		const item = data instanceof FormData ? Object.fromEntries(data) : (data ?? {});
		this.id = item.id;
		this.name = item.name ?? '';
		this.type = item.type ?? 'website';
		this.target = item.target ?? '';
		this.token = item.token ?? '';
		this.interval = toInterval(item.interval);
		this.enabled = toBool(item.enabled);
		this.zones = toZones(item.zones);
		this.tags = toZones(item.tags);
		this.escalationPolicy = item.escalation_policy ?? item.escalationPolicy ?? '';
		const cf = configFields(item);
		this.keyword = cf.keyword;
		this.keywordMode = cf.keywordMode;
		this.method = cf.method;
		this.headers = cf.headers;
		this.expectedStatus = cf.expectedStatus;
		this.followRedirects = cf.followRedirects;
		this.timeoutSecs = cf.timeoutSecs;
		this.maintenanceWindows = cf.maintenanceWindows;
		this.latencyThresholdMs = cf.latencyThresholdMs;
		this.dnsRecordType = cf.dnsRecordType;
		this.dnsExpectedValue = cf.dnsExpectedValue;
		this.dnsResolver = cf.dnsResolver;
		// A disabled monitor isn't scheduled (worker/evaluator filter enabled=true),
		// so the backend never writes 'paused' — its stored status just goes stale.
		// Surface "paused" from `enabled` here so every view reflects it consistently.
		this.status = this.enabled ? (item.status ?? 'pending') : 'paused';
		this.consensusZones = toZones(item.consensus_zones ?? item.consensusZones);
		this.consensusFresh = toZones(item.consensus_fresh ?? item.consensusFresh);
		this.lastChecked = item.last_checked ?? item.lastChecked ?? null;
		this.uptime24h = toNumberOrNull(item.uptime24h);
		this.certExpiresAt = item.cert_expires_at ?? item.certExpiresAt ?? null;
		this.domainExpiresAt = item.domain_expires_at ?? item.domainExpiresAt ?? null;
		this.lastDowntime = item.lastDowntime ?? item.last_downtime ?? null;
	}
}

export class MonitorNewItem {
	name: string;
	type: MonitorType;
	target: string;
	interval: number;
	enabled: boolean;
	zones: string[];
	tags: string[];
	escalationPolicy: string;
	keyword: string;
	keywordMode: string;
	method: string;
	headers: string;
	expectedStatus: number | null;
	followRedirects: boolean;
	timeoutSecs: number | null;
	maintenanceWindows: MaintenanceWindow[];
	latencyThresholdMs: number | null;
	dnsRecordType: string;
	dnsExpectedValue: string;
	dnsResolver: string;

	constructor(data: FormData | RecordModel | any = null) {
		const fromForm = data instanceof FormData;
		const item = fromForm ? Object.fromEntries(data) : (data ?? {});
		this.name = item.name ?? '';
		this.type = item.type ?? 'website';
		this.target = item.target ?? '';
		this.interval = item.interval !== undefined ? toInterval(item.interval) : 60;
		this.enabled = fromForm ? toBool(item.enabled) : (item.enabled ?? true);
		this.zones = toZones(item.zones);
		this.tags = toZones(item.tags);
		this.escalationPolicy = item.escalation_policy ?? item.escalationPolicy ?? '';
		const cf = configFields(item);
		this.keyword = cf.keyword;
		this.keywordMode = cf.keywordMode;
		this.method = cf.method;
		this.headers = cf.headers;
		this.expectedStatus = cf.expectedStatus;
		this.followRedirects = cf.followRedirects;
		this.timeoutSecs = cf.timeoutSecs;
		this.maintenanceWindows = cf.maintenanceWindows;
		this.latencyThresholdMs = cf.latencyThresholdMs;
		this.dnsRecordType = cf.dnsRecordType;
		this.dnsExpectedValue = cf.dnsExpectedValue;
		this.dnsResolver = cf.dnsResolver;
	}
}

export class MonitorItemValidator extends BaseEntityValidator {
	constructor(item: MonitorItem | MonitorNewItem) {
		super(schema, item);
	}
}

export class MonitorsList extends BaseEntityList<MonitorItem> {
	constructor(data: MonitorsList);
	constructor(data: ListResult<RecordModel> | any) {
		// Guard against the uninitialised shared store (`{}`), which has no
		// `.items`: on the monitor detail page `store.edit()` updates this store
		// before the list page has ever populated it — without the guard
		// `data.items.map` throws, the modal's submit rejects, and it never closes.
		super(
			data instanceof MonitorsList
				? data.items
				: (data?.items ?? []).map((it: any) => new MonitorItem(it)),
			data instanceof MonitorsList
				? new PaginationData(data.pagination)
				: new PaginationData(data ?? {})
		);
	}

	add(item: any) {
		super.add(new MonitorItem(item));
		return this;
	}

	edit(item: any) {
		super.edit(new MonitorItem(item));
		return this;
	}

	remove(id: EntityIdType) {
		super.remove(id);
		return this;
	}
}

File: corelib/models/Monitor.go
package models

import "time"

// MinIntervalSeconds is the minimum allowed check interval for real monitors
// (#319). Enforced at create/edit by the schema (monitors.interval min) and the
// form; EffectiveInterval clamps to it defensively so a legacy row created
// under the old 5s floor can't probe faster than this. A single knob if we ever
// want per-tier floors. (Anonymous trials floor much higher, at 300s.)
const MinIntervalSeconds = 30

// DefaultIntervalSeconds is the cadence assumed for a monitor with no interval
// stored (the field is optional, so 0 means "unset", not "as fast as possible").
const DefaultIntervalSeconds = 60

// EffectiveInterval is the cadence the system actually runs a monitor at: unset
// falls back to the default, and anything under the floor is clamped up.
//
// Every component that reasons about cadence MUST go through this. The worker
// schedules on it and the evaluator derives its freshness window from it; if
// only one of them clamped, a legacy sub-floor monitor would be probed every 30s
// while being judged against a window built from its raw 5s interval, so its
// checks would read as stale and it could never reach a verdict.
func EffectiveInterval(raw int) int {
	if raw <= 0 {
		return DefaultIntervalSeconds
	}
	if raw < MinIntervalSeconds {
		return MinIntervalSeconds
	}
	return raw
}

// Monitor is a sensor definition as stored in the PocketBase `monitors` collection.
type Monitor struct {
	Id       string   `json:"id"`
	User     string   `json:"user"`
	Name     string   `json:"name"`
	Type     string   `json:"type"`
	Target   string   `json:"target"`
	Interval int      `json:"interval"`
	Zones    []string `json:"zones"`
	Enabled  bool     `json:"enabled"`
	Status   string   `json:"status"`
	// ConsensusZones / ConsensusFresh record what the LAST evaluation actually
	// voted with: the zones that took part, and the subset of those that had a
	// fresh check (#328). Comma-separated, written by the evaluator only when they
	// change.
	//
	// They exist because the consequence of a zone going quiet is otherwise
	// invisible: a monitor pinned to two zones silently falls back to the
	// single-zone rule, and the UI cannot re-derive that honestly — freshness is
	// computed from the monitor's effective interval inside the evaluator, so any
	// second guess drifts from the decision that was actually made.
	ConsensusZones string `json:"consensus_zones"`
	ConsensusFresh string `json:"consensus_fresh"`
	// LastChecked is when the monitor was last probed or — for heartbeat monitors
	// — last checked in. Heartbeat evaluation uses it for staleness.
	LastChecked string `json:"last_checked"`
	// Token is the unguessable check-in token for heartbeat monitors (empty for
	// probed types); the public /ping/{token} endpoint records a check-in.
	Token string `json:"token"`
	// CertExpiresAt is when the target's TLS certificate expires (HTTPS website
	// monitors only), captured during the check. Empty if unknown/not HTTPS.
	CertExpiresAt string `json:"cert_expires_at"`
	// DomainExpiresAt is when the target's DOMAIN registration lapses (distinct
	// from the TLS cert above), from an infrequent, cached RDAP/WHOIS lookup the
	// evaluator runs — not the per-check probe. Empty if unknown (no data, a
	// non-domain target, or a rate-limited lookup). DomainCheckedAt is when that
	// lookup last ran, used to keep it infrequent.
	DomainExpiresAt string `json:"domain_expires_at"`
	DomainCheckedAt string `json:"domain_checked_at"`
	// EscalationPolicy is the id of the escalation_policies record to drive
	// paging for this monitor, or "" for the default (all channels, once).
	EscalationPolicy string `json:"escalation_policy"`
	// Config holds optional per-monitor check options (JSON field).
	Config MonitorConfig `json:"config"`
}

// MonitorConfig is the per-monitor `config` JSON blob. Optional check tuning
// that not every monitor sets. Every field has a safe zero-value default so an
// unset config behaves exactly like the original hardcoded check.
type MonitorConfig struct {
	// Keyword + KeywordMode assert on the response body: "contains" requires the
	// keyword to be present, "absent" requires it to be missing. Empty mode =
	// no body assertion.
	Keyword     string `json:"keyword,omitempty"`
	KeywordMode string `json:"keywordMode,omitempty"`
	// Method is the HTTP method (GET/HEAD/POST); empty = GET.
	Method string `json:"method,omitempty"`
	// Headers are extra request headers to send (override our defaults by name).
	Headers map[string]string `json:"headers,omitempty"`
	// ExpectedStatus, when non-zero, requires an exact status code for "up"
	// (otherwise the default 200–399 rule applies).
	ExpectedStatus int `json:"expectedStatus,omitempty"`
	// FollowRedirects toggles 3xx following. nil = default (follow); false = stop
	// at the first response and evaluate it as-is.
	FollowRedirects *bool `json:"followRedirects,omitempty"`
	// TimeoutSecs overrides the per-check timeout in seconds; 0 = default.
	TimeoutSecs int `json:"timeoutSecs,omitempty"`
	// MaintenanceWindows suppress alerts during planned work: while one is active
	// the monitor still runs checks, but no down/recovery alerts fire.
	MaintenanceWindows []MaintenanceWindow `json:"maintenanceWindows,omitempty"`
	// LatencyThresholdMs, when > 0, alerts on slowness: the evaluator warns when
	// the recent p95 response time exceeds this (with flap damping), separately
	// from up/down. 0 = disabled.
	LatencyThresholdMs int `json:"latencyThresholdMs,omitempty"`
	// DNS monitor options. DNSRecordType is A/AAAA/CNAME/MX/TXT/NS (empty = A).
	// DNSExpectedValue, when set, requires a resolved record to contain it.
	// DNSResolver, when set, queries that DNS server (host or host:port) instead
	// of the system resolver.
	DNSRecordType    string `json:"dnsRecordType,omitempty"`
	DNSExpectedValue string `json:"dnsExpectedValue,omitempty"`
	DNSResolver      string `json:"dnsResolver,omitempty"`
}

// MaintenanceWindow is a one-off planned-work interval (RFC3339 timestamps).
// Alerts are suppressed while now is within [Start, End).
type MaintenanceWindow struct {
	Start string `json:"start"`
	End   string `json:"end"`
}

// EscalationPolicy is an ordered set of levels that fire over time while an
// incident stays open and unacknowledged.
type EscalationPolicy struct {
	Id    string           `json:"id"`
	Name  string           `json:"name"`
	Steps []EscalationStep `json:"steps"`
}

// EscalationStep is one level: after AfterMinutes since the incident opened,
// notify these channel ids (if still open + unacknowledged).
type EscalationStep struct {
	AfterMinutes int      `json:"after_minutes"`
	Channels     []string `json:"channels"`
}

// CheckResult is the outcome of running a single check against a monitor's target.
type CheckResult struct {
	Up         bool
	StatusCode int
	ResponseMs int
	DnsMs      int
	ConnectMs  int
	TlsMs      int
	TtfbMs     int
	Error      string
	// CertExpiresAt is the target's TLS certificate expiry, captured from the
	// handshake during an HTTPS check. Zero when there's no TLS/cert.
	CertExpiresAt time.Time
	// Blocked marks a result as rate-limited / bot-blocked (e.g. HTTP 429/403):
	// the target refused to let us judge it, which is neither "up" nor a real
	// "down". The evaluator treats it as a neutral abstention (no incident, no
	// alert). Not persisted — derived downstream from StatusCode.
	Blocked bool
	// RetryAfterSecs is the target's requested backoff (from the Retry-After
	// header) when Blocked; 0 if absent. Used to reschedule the next check
	// instead of hammering. Not persisted.
	RetryAfterSecs int
	// RedirectCount is how many redirects were followed to reach the final
	// response (0 when none). FinalURL is the URL of that final response — set
	// only when RedirectCount > 0. Transparency only (#112); the final response
	// still decides up/down.
	RedirectCount int
	FinalURL      string
}
